cip/1.accepted/CIP2016-12-14-Constraint-syntax.adoc

```diff
@@ -0,0 +1,285 @@
+
+==== Syntax
+
+.Grammar definition for property existence constraint:
+[source, ebnf]
+----
+existence-constraint = "exists", "(", property-expression, ")" ;
+----
+
+.Example of node property existence constraint:
+[source, cypher]
+----
+CREATE CONSTRAINT colors_must_have_rgb
+FOR (c:Color)
+REQUIRE exists(c.rgb)
+----
+
+.Example of relationship property existence constraint:
+[source, cypher]
+----
+CREATE CONSTRAINT rates_have_quality
+FOR ()-[l:RATED]-()
+REQUIRE exists(l.rating)
+----
+
+==== Semantics
+
+The domain of a node property existence constraint are all nodes with the specified label.
+Similarly, the domain of a relationship property existence constraint are all relationship with the specified type.
+
+Property existence constraints mandates that the value of the specified property exists (is non-null) for all entities in the domain.
+
+===== Example
+
+Consider the graph containing `:Color` nodes.
+Each color has an integral RGB value representation in a property `rgb`.
+Users may lookup color nodes to extract their RGB values for application processing.
+Users may also add new color nodes to the graph.
+
+Suppose the query that looks up the RGB value of a color with a given name looks like this:
+
+[source, cypher]
+----
+MATCH (c:Color {name: $name})
+WHERE exists(c.rgb)
+RETURN c.rgb
+----
+
+The `WHERE` clause protects the application from receiving `null` values back for user-defined colors where the RGB values have not been specified correctly.
+It may however be eliminated by the introduction of a node property existence constraint:
+
+[source, cypher]
+----
+CREATE CONSTRAINT colors_must_have_rgb
+FOR (c:Color)
+REQUIRE exists(c.rgb)
+----
+
+Any updating statement that would create a `:Color` node without specifying a `rgb` property for it would now fail.
+
+=== Interaction with existing features
+
+The main interaction between the constraints and the rest of the language happens during updating statements.
+Existing constraints will cause certain updating statements to fail; in fact, that's the main purpose.
+
+=== Alternatives
+
+Plenty of alternative syntaxes have been discussed:
+
+* `GIVEN`, `CONSTRAIN`, `ASSERT` instead of `FOR`
+* `ASSERT`, `ENFORCE`, `IMPLIES` instead of `REQUIRE`
+
+The use of existing expression to express uniqueness, instead of using a new keyword `UNIQUE`, on the form:
+----
+FOR (p:Person), (q:Person)
+REQUIRE p.email <> q.email AND p <> q
+----
+which quickly becomes unwieldy for multiple properties.
+
+== What others do
+
+In SQL, the following constraints exist (http://www.w3schools.com/sql/sql_constraints.asp):
+
+* `NOT NULL` - Indicates that a column cannot store NULL value
+* `UNIQUE` - Ensures that each row for a column must have a unique value
+* `PRIMARY KEY` - A combination of a `NOT NULL` and `UNIQUE`. Ensures that a column (or combination of two or more columns) have a unique identity which helps to find a particular record in a table more easily and quickly
```

Comment:
"which helps to find a particular record in a table more easily and quickly"
->
", reducing the resources required to locate a specific record in a table"